src/core/txs/validate_tx.c

```diff
@@ -0,0 +1,146 @@
+/**
+ * @file validate_tx.c
+ * @author Nathan Faber nfaber@olin.edu
+ * @brief 
+ * @version 0.1
+ * @date 2022-03-20
+ * 
+ * @copyright Copyright (c) 2022
+ * 
+ */
+#include "base_tx.h"
+#include "utxo_pool.h"
+#include "mempool.h"
+#include "constants.h"
+#include "mbedtls/ecdsa.h"
+#include "sign_tx.h"
+#include "create_block.h"
+
+void create_blank_sig_txhash(unsigned char *blank_hash, Transaction *tx){
+
+  Transaction *blank_sig_tx = malloc(sizeof(Transaction));
+  memcpy(blank_sig_tx, tx, sizeof(Transaction));
+
+  blank_sig_tx->inputs = malloc(tx->num_inputs*sizeof(Input));
+  blank_sig_tx->outputs = malloc(tx->num_outputs*sizeof(Output));
+
+  memcpy(blank_sig_tx->inputs, tx->inputs, tx->num_inputs*sizeof(Input));
+  memcpy(blank_sig_tx->outputs, tx->outputs, tx->num_outputs*sizeof(Output));
+  
+  for(unsigned int i = 0; i < blank_sig_tx->num_inputs; i++){
+    blank_sig_tx->inputs[i].sig_len = 0;
+    memset(blank_sig_tx->inputs[i].signature, 0, SIGNATURE_LEN);
+  }
+  // need to make hash where the tx has signatures that are 0's
+  hash_tx(blank_hash, blank_sig_tx);
+  free_tx(blank_sig_tx);
+}
+
+int check_input_unlockable(Input *input, unsigned char *blank_tx_hash){
+  unsigned char test_pub_key_hash[PUB_KEY_HASH_LEN];
+  mbedtls_ecdsa_context new_key;
+  UTXO *old_utxo;
+  //unsigned char blank_tx_hash[TX_HASH_LEN];
+  // create_blank_sig_txhash(blank_tx_hash, tx);
+
+  build_ctx_from_public(&new_key, input->pub_key);
+  if(validate_sig(input->signature, input->sig_len, blank_tx_hash,
+                  TX_HASH_LEN, &new_key) != 0){
+                    return 1;
+                  }
+  return 0;
+}
+
+int validate_inputs_vs_outputs(Transaction *tx){
+  // This functionality is extremely similar to calc_tx_fees....
+  UTXO *old_utxo;
+  unsigned int total_in = 0;
+  unsigned int total_out = 0;
+  for(unsigned int i = 0; i < tx->num_inputs; i++){
+    // Make sure the referenced utxo exists and the pubilc key matches the hash
+    old_utxo = utxo_pool_find(tx->inputs[i].prev_tx_id, tx->inputs[i].prev_utxo_output);
+    if(old_utxo == NULL){
+      return 1;
+    }
+    total_in += old_utxo->amt;
+  }
+  for(unsigned int i = 0; i < tx->num_outputs; i++){
+    total_out += tx->outputs[i].amt;
+  }
+
+  if(total_in < total_out){
+    return 2;
+  }
+  return 0;
+}
+
+
+int validate_input_not_in_mempool(Input *input){
+  Transaction *mempool_tx;
+  mempool_tx = mempool_find(input->prev_tx_id);
+  if(mempool_tx != NULL){
+    return 1;
+  }
+}
+
+int validate_input_matches_utxopool(Input *input){
+  mbedtls_ecdsa_context new_key;
+  unsigned char test_pub_key_hash[PUB_KEY_HASH_LEN];
+  build_ctx_from_public(&new_key, input->pub_key);
+  hash_pub_key(test_pub_key_hash, &new_key);
+  UTXO *old_utxo = utxo_pool_find(input->prev_tx_id, input->prev_utxo_output);
+  if(old_utxo == NULL){
+    return 1;
+  }
+  if(memcmp(old_utxo->public_key_hash, test_pub_key_hash, PUB_KEY_HASH_LEN) != 0){
+    return 2;
+  }
+  return 0;
```

Comment:
I think we were also planning on checking to see if the spent flag was set on the found UTXO, although I'm not sure if we want to follow through with that.

Desired functionality would be to mark a tx invalid if the UTXO it references was already spent by a tx in the mempool. This does raise some questions about user functionality, since it means you couldn't send a second tx that references a UTXO if you already sent one that spent it. Also raises issues during block validation time, since the tx in the block is likely (although not necessarily) also in the mempool.

IIRC the purpose of this feature was to avoid additional overhead when selecting things to mine - specifically, we need to make sure we don't have two transactions that reference the same UTXO. We would also need to clean the mempool of transactions that become invalid after one of their UTXOs are spent.

Given the issues outlined above, my vote would be to get rid of this in validation, and handle it in a new PR  down the line. This would have to add to block validation and block handling, and will probably be non-trivial (some design decision to make) and isn't critical for our demo, so I would also vote to do this later.

Curious to hear your thoughts.

Reply:
I think you bring up a good point in checking for double spend in block verification. We don't do it right now and we need to.

However, I disagree on requesting txs. We can't trust a node to send valid transactions, so we'll have to validate everything anyways. Moreover, we couldn't trust any data the other node sends that isn't already in the block so there isn't much point. It would be much more efficient to just check the block for double spend.

Here's my proposal:
- Update block verification to check for double spend
- Update mine transaction collection to avoid double spend
- Update handle block to clean mempool
- Remove spent flag from UTXOs (not wallet, which is separate)

This doesn't matter for the demo or this PR, so I would just say we open an issue and do it later.

(I will double check what bitcoin does when I get the chance)